### ISSUE TITLE
Git work with submodules that have been checked out by other means

### ISSUE THREAD

**Issue by danielr1996** [isomorphic-git/isomorphic-git#1647]
I'd like to propose a feature as well as a solution: 

Theoretically to reach 100% git compatibility this library should handle setup as well as working with checked out submodules. But for now lets focus on working with submodules that have been checked out by other means. 

As far as my understanding of submodules go to git dir is not stored as `.git`, instead `.git` is a file that points to the actual git directory. So is wrote this little workarround in Deno which works for me: 
```
let gitdir = '.git';
const stat =await Deno.stat(gitdir)
if (stat.isFile) {
    const dotgit = await Deno.readTextFile(gitdir)
    gitdir = path.join(Deno.cwd(), dotgit.replace('gitdir: ','').replace('\r','').replace('\n',''))
}

//@ts-ignore
const tagOid = await git.resolveRef({fs, dir: Deno.cwd(), gitdir, ref: 'main'});
```

Is this really all there is to support working with submodules? Or are there edge cases that I didn't consider. 
If it's really this simple I might create a PR if I *find the time™*

**Comment by jcubic** [comment]
Unfortunately I have no idea, @mojavelinux maybe you know?

**Comment by mojavelinux** [comment]
Yes, that's essentially it. If `.git` is a file rather than a directory, it points to a linked / embedded repository under the `.git` folder of the main repository. This linked repository can either be a linked worktree or a (sub)module. A linked worktree will contain the `commondir` and `gitdir` files, whereas a module will not.

Here's code I put together to test this logic:

```js
const fs = require('fs')
const { promises: fsp } = fs
const git = require('isomorphic-git')
const ospath = require('path')

;(async () => {
  let repo = { fs, dir: 'js' }
  const dotgit = ospath.join(repo.dir, '.git')
  const dotgitStat = await fsp.stat(dotgit).catch(() => ({ isFile: () => false, isDirectory: () => false }))
  if (dotgitStat.isFile()) repo = await resolveRepositoryFromSubmodule({ ...repo, gitdir: dotgit })
  console.log(await git.resolveRef({ ...repo, ref: 'HEAD' }))
})()

function resolveRepositoryFromSubmodule (repo) {
  return fsp
    .readFile(repo.gitdir, 'utf8')
    .then((contents) => contents.trimRight().substr(8))
    .then((submoduleGitdir) => {
      const gitdir = ospath.join(repo.dir, submoduleGitdir)
      return Object.assign(repo, { gitdir, submoduleName: ospath.basename(submoduleGitdir) })
    })
}
```

We could take it a step further and add support for a linked worktree as well:

```js
const fs = require('fs')
const { promises: fsp } = fs
const git = require('isomorphic-git')
const ospath = require('path')

;(async () => {
  let repo = { fs, dir: 'js' }
  const dotgit = ospath.join(repo.dir, '.git')
  const dotgitStat = await fsp.stat(dotgit).catch(() => ({ isFile: () => false, isDirectory: () => false }))
  if (dotgitStat.isFile()) repo = await resolveRepositoryFromDotGit({ ...repo, gitdir: dotgit })
  console.log(await git.resolveRef({ ...repo, ref: 'HEAD' }))
  console.log(await git.listFiles({ ...repo }))
})()

function resolveRepositoryFromDotGit (repo) {
  return fsp
    .readFile(repo.gitdir, 'utf8')
    .then((contents) => contents.trimRight().substr(8))
    .then((gitdir) => {
      const name = ospath.basename(gitdir)
      if (ospath.basename(ospath.dirname(gitdir)) === 'modules') {
        gitdir = ospath.join(repo.dir, gitdir)
        return Object.assign(repo, { gitdir, submoduleName: name })
      }
      return fsp.readFile(ospath.join(gitdir, 'commondir'), 'utf8').then(
        (contents) => {
          gitdir = ospath.join(gitdir, contents.trimRight())
          const dir = ospath.basename(gitdir) === '.git' ? ospath.dirname(gitdir) : gitdir
          return Object.assign(repo, { dir, gitdir, worktreeName: name })
        },
        () => repo
      )
    })
}
```

I'm not really sure where this belongs in the API. A simple approach would be to add a function such as resolveRepository that will find the main repository given the repo parameters you provided. In other words, follow a worktree or module to the main repository. But what you end up doing with that information depends on how you are using the subrepository type.

**Comment by mojavelinux** [comment]
@danielr1996 Are you suggesting that the calling code knows that the folder is a submodule (or perhaps would like isomorphic-git to tell you that), then it will work with the submodule as a separate repository?

**Comment by danielr1996** [comment]
@mojavelinux sorry for the late answer, as I got it working for me I kinda forgot about that issue. 

Well I think there are two possible solutions, the one I would expected as a user would be that the library resolves the real git directory and I, as a end user don't need to care wether I use a normal git repo, submodule or linked worktree. 

However I understand that this might be and edge case and resolving the git dir might be maintenance overhead that could be avoided. For that case I would at least hope to find an explanation in the documentation that the library by default only works with normal git repos, but can easiliy be taught to work with submodules or a linked working tree by doing the following things <insert my example of resolving the real dir>

I think the second solution would be fine, as it explains the behaviour and provides a good solution that I had to find be digging around and experimenting.

**Comment by kuncevic** [comment]
Any thoughts if submodule support may land at some point?

**Comment by jcubic** [comment]
@kuncevic I don't think anyone is working on this. This project is community driven which means that any changes to the project are done by individual contributors. There is no main author that writes all the code, he left the project. Do you want to contribute?

**Comment by kuncevic** [comment]
> @kuncevic I don't think anyone is working on this. This project is community driven which means that any changes to the project are done by individual contributors. There is no main author that writes all the code, he left the project. Do you want to contribute?

yeah I c, thank you for clarify. Speaking contribution let me see, had to assess that.

**Comment by unilynx** [comment]
I also ran into this one today. My lazy quick workaround was changing

```
  const branch = await git.currentBranch({ fs, dir }) || 'HEAD';
```

to

```
  const gitdir = (await fs.stat(path.join(dir, ".git"))).isDirectory() ?
    path.join(dir, ".git") :
    (await fs.readFile(path.join(dir, ".git"), "utf8")).trim().substring(8);

  const branch = await git.currentBranch({ fs, dir, gitdir }) || 'HEAD';
```

**Comment by scolladon** [comment]
### What should I do if I want to `diff --name-status` a repository and its submodules ?

I assume I would need to use the `walk` API, but I don't know how to detect when the diff is from a submodule and I don't know either how to introspect the diff on that submodules as well.

Starting from this code snippet (adapted from [here](https://isomorphic-git.org/docs/en/snippets#git-diff-name-status-commithash1-commithash2)), how should I do ?
```js
async function getFileStateChanges(commitHash1, commitHash2, dir) {
    return git.walk({
      fs,
      dir,
      trees: [git.TREE({ ref: commitHash1 }), git.TREE({ ref: commitHash2 })],
      map: async function(filepath, trees) {
        if (filepath === '.') {
          return
        }
        for (const tree of trees.filter(Boolean)) {
          const type = await tree!.type()
		  // This is where I should include commit type as well ?
          // And what should I do then to instrospect the submodules and diff ? another `walk` ?
          if (type !=='blob') {
            return
          }
        }
  
        const [fromOID, toOID] = await Promise.all(trees.map(tree => tree?.oid()))
        if (fromOID === toOID) {
          return
        }
  
        let type = 'M'
        if (fromOID === undefined) {
          type = 'A'
        } else if (toOID === undefined) {
          type = 'D'
        }

        return `${type}\t${filepath}`
      },
    })
  }
```

**Comment by jcubic** [comment]
This looks like a question that is someone related to the issue. I would ask another question. Note that I don't know the answer to this question.

**Comment by scolladon** [comment]
@jcubic I have created a question [here](https://stackoverflow.com/questions/77782643/how-to-diff-name-status-a-repository-and-its-submodules)

**Comment by sdarwin** [comment]
> But for now lets focus on working with submodules that have been checked out by other means.

It would be really helpful to have this feature!

**Comment by jcubic** [comment]
@sdarwin do you want to contribute?

**Comment by sdarwin** [comment]
> do you want to contribute?

maybe we'll be able to contribute but there is no timeline so if someone else does first that would be fine.
